### PR TITLE
Add readline support to sqlite

### PR DESCRIPTION
**Pull request by jerowe**
Hello,

I'm not sure how to go about specifying readline.

In dependencies I have : readline 6.3

And in the travis test I get - 

AssertionError: 'sqlite 3.13.0 readline 6.3_0'

Thanks!

Best,
Jillian

### CODE REVIEW

File: recipe/meta.yaml [hunk 1]
```diff
@@ -23,11 +23,13 @@ build:
     - vc9    # [win and py27]
     - vc10   # [win and py34]
     - vc14   # [win and py35]
+    - readline 6.3
```
Comment (jakirkham) [inline recipe/meta.yaml]: Please drop this line.

File: recipe/meta.yaml [hunk 2]
```diff
@@ -23,11 +23,13 @@ build:
...
 
 requirements:
   build:
     # python is for lining up vc feature (runtime)
     - python      # [win]
+    - readline 6.3
```
Comment (jakirkham) [inline recipe/meta.yaml]: Please change to `6.2` to match our [global pinnings]( https://github.com/conda-forge/conda-forge.github.io/blob/70036d169783f7d6844bfc1ff1b360dba5adfa9f/scripts/pin_the_slow_way.py#L48 ).

File: recipe/meta.yaml [hunk 3]
```diff
@@ -23,11 +23,13 @@ build:
...
```
Comment (jakirkham) [inline recipe/meta.yaml]: Please add a `run` section to `requirements` that contains `readline 6.2`.

File: recipe/build.sh [hunk 4]
```diff
@@ -4,11 +4,12 @@ if [ `uname -m` == ppc64le ]; then
     B="--build=ppc64le-linux"
 fi
 
+#--disable-readline \
+
```
Comment (jakirkham) [inline recipe/build.sh]: Guessing we can drop these two lines.

**Comment by jerowe** [comment]
I am generally interested in the conda forge build process, but haven't been able to find much about it.

Why are there 3 different CI services? What are they doing? Could you point me towards some docs?

**Comment by jerowe** [comment]
We've moved onto different failures. Progress!!

```
+ source /opt/conda/bin/activate /feedstock_root/build_artefacts/recipe_root_1474276388947/_t_env
+ /bin/bash -x -e /feedstock_root/build_artefacts/recipe_root_1474276388947/test_tmp/run_test.sh
+ sqlite3 --version
3.13.0 2016-05-18 10:57:30 fc49f556e48970561d7ab6a2f24fdd7d9eb81ff2
TEST END: sqlite-3.13.0-1
INFO:/opt/conda/lib/python3.5/site-packages/conda_build/config.py:--keep-old-work flag not specified.  Removing source and build files.

# If you want to upload this package to anaconda.org later, type:
#
# $ anaconda upload /feedstock_root/build_artefacts/linux-64/sqlite-3.13.0-1.tar.bz2
#
# To have conda build upload to anaconda.org automatically, use
# $ conda config --set anaconda_upload yes

Using Anaconda API: https://api.anaconda.org
Traceback (most recent call last):
  File "/feedstock_root/ci_support/upload_or_check_non_existence.py", line 124, in <module>
    main()
  File "/feedstock_root/ci_support/upload_or_check_non_existence.py", line 104, in main
    exists = built_distribution_already_exists(cli, meta, owner)
  File "/feedstock_root/ci_support/upload_or_check_non_existence.py", line 24, in built_distribution_already_exists
    fname = bldpkg_path(meta)
TypeError: bldpkg_path() missing 1 required positional argument: 'config'

./ci_support/run_docker_build.sh returned exit code 1
```

**Comment by jakirkham** [comment]
Yup, this error will be easy to fix. Basically we have to re-render, which had to do anyways as we have added Python. To do this, [install `conda-smithy`]( https://github.com/conda-forge/conda-smithy/blob/6a55ee6b813a0516a13d341b4af3ea3d89754ea6/README.md#installation ) and [re-render the feedstock]( https://github.com/conda-forge/conda-smithy/blob/6a55ee6b813a0516a13d341b4af3ea3d89754ea6/README.md#re-rendering-an-existing-feedstock ).

**Comment by jakirkham** [comment]
> Why are there 3 different CI services? What are they doing? Could you point me towards some docs?

Docs are pretty sparse ATM. Basically we need a way to build on the 3 major OSes. So we have Mac (Travis CI), Linux (CircleCI), and Windows (AppVeyor).

**Comment by jerowe** [comment]
@jakirkham, thanks for answering my questions!

I didn't forget about this, but a lot of other tasks came up this week.

The windows build failed. I had such high hopes! Should I just skip readline for windows? How do I deal with the pin then?
```
Error: Package missing in current win-32 channels: 
  - readline 6.2
Command exited with code 1
```

File: appveyor.yml [hunk 5]
```diff
@@ -20,22 +20,8 @@ environment:
...
-      CONDA_PY: 34
-
-    - TARGET_ARCH: x64
-      CONDA_PY: 34
-
-    - TARGET_ARCH: x86
-      CONDA_PY: 35
-
-    - TARGET_ARCH: x64
-      CONDA_PY: 35
```
Comment (jakirkham) [inline appveyor.yml]: So this is off. Not sure why the re-rendering did this too us. Could you please try re-rendering again to see if it goes away? If it doesn't go away, it would be great to get some info about your setup.

**Comment by jerowe** [comment]
@jakirkham, could you take a look?

File: appveyor.yml (hunk 5 above)
Comment (jerowe) [inline appveyor.yml]: I just saw this. let me try again.

**Comment by jakirkham** [comment]
Basically fine, but the re-rendering went wrong.

We have occasionally seen this issue ( https://github.com/conda-forge/conda-smithy/issues/223 ), but it has been very difficult to narrow down. Even if we aren't able to fix the re-rendering, but do find a reproducer that could prove incredibly useful for fixing that bug.

**Comment by jerowe** [comment]
```
No circle token.  Create a token at https://circleci.com/account/api and
put it in ~/.conda-smithy/circle.token
No appveyor token. Create a token at https://ci.appveyor.com/api-token and
Put one in ~/.conda-smithy/appveyor.token
No anaconda token. Create a token via
  anaconda auth --create --name conda-smithy --scopes "repos conda api"
and put it in ~/.conda-smithy/anaconda.token
Using Anaconda API: https://api.anaconda.org
Fetching package metadata ...........

CI support files regenerated. These need to be pushed to github!
```

```
Current conda install:

             platform : linux-64
        conda version : 4.1.11
    conda-env version : 2.5.2
  conda-build version : 2.0.1
       python version : 3.5.2.final.0
     requests version : 2.9.1
     root environment : /home/jillian/.local/easybuild/software/anaconda/3-4.0.0  (writable)
  default environment : /home/jillian/.local/easybuild/software/anaconda/3-4.0.0
     envs directories : /home/jillian/.local/easybuild/software/anaconda/3-4.0.0/envs
        package cache : /home/jillian/.local/easybuild/software/anaconda/3-4.0.0/pkgs
         channel URLs : https://conda.anaconda.org/t/<TOKEN>/r/linux-64/
                        https://conda.anaconda.org/t/<TOKEN>/r/noarch/
                        https://conda.anaconda.org/t/<TOKEN>/bioconda/linux-64/
                        https://conda.anaconda.org/t/<TOKEN>/bioconda/noarch/
                        https://repo.continuum.io/pkgs/free/linux-64/
                        https://repo.continuum.io/pkgs/free/noarch/
                        https://repo.continuum.io/pkgs/pro/linux-64/
                        https://repo.continuum.io/pkgs/pro/noarch/
          config file : /home/jillian/.condarc
         offline mode : False
    is foreign system : False
```

File: appveyor.yml (hunk 5 above)
Comment (jakirkham) [inline appveyor.yml]: Ok, so this just worked the second time. The sporadic nature of this issue has made things difficult. If you have any details about the first re-rendering that you can share, that would be very helpful (e.g. packages in your environment, log, etc.).

**Comment by jakirkham** [comment]
Seems to be picking up the [system `libedit` on Mac]( https://travis-ci.org/conda-forge/sqlite-feedstock/builds/161975034#L515 ) instead of `readline`. Is there some option we can use to specify where to look for `readline`? If not, we can try things like [this]( https://github.com/conda-forge/htop-feedstock/blob/e4e54dba8a093e0082d2771063c57c7939360df9/recipe/build.sh#L3 ).

File: appveyor.yml (hunk 5 above)
Comment (jerowe) [inline appveyor.yml]: That is odd. 

conda list
```
# packages in environment at /home/jillian/.local/easybuild/software/anaconda/3-4.0.0:
#
Using Anaconda API: https://api.anaconda.org
alabaster                 0.7.7                    py35_0  
anaconda-client           1.5.1                    py35_0  
anaconda                  custom                   py35_0  
anaconda-navigator        1.1.0                    py35_0  
argcomplete               1.0.0                    py35_1  
astropy                   1.1.2               np110py35_0  
babel                     2.2.0                    py35_0  
beautifulsoup4            4.4.1                    py35_0  
BeeSQL                    0.1                       <pip>
biopython                 1.68                     py35_0    bioconda
hmmer                     3.1b2                         1    bioconda
mmtf-python               1.0.2                    py35_0    bioconda
bitarray                  0.8.1                    py35_0  
blaze                     0.9.1                    py35_0  
bokeh                     0.11.1                   py35_0  
boto                      2.39.0                   py35_0  
bottleneck                1.0.0               np110py35_0  
cffi                      1.5.2                    py35_0  
chest                     0.2.3                    py35_0  
cloudpickle               0.1.1                    py35_0  
clyent                    1.2.1                    py35_0  
colorama                  0.3.7                    py35_0  
conda-build               2.0.1                    py35_0  
click                     6.6                      py35_0    conda-forge
conda                     4.1.11                   py35_1    conda-forge
conda-build-all           1.0.0                    py35_0    conda-forge
conda-env                 2.5.2                    py35_0    conda-forge
conda-smithy              1.2.0                    py35_1    conda-forge
filelock                  2.0.6                    py35_0    conda-forge
gitdb                     0.6.4                    py35_0    conda-forge
gitpython                 2.0.8                    py35_0    conda-forge
mock                      2.0.0                    py35_0    conda-forge
pbr                       1.10.0                   py35_0    conda-forge
pygithub                  1.27.1                   py35_0    conda-forge
ruamel.yaml               0.11.11                  py35_1    conda-forge
smmap                     0.9.0                    py35_0    conda-forge
conda-manager             0.3.1                    py35_0  
configobj                 5.0.6                    py35_0  
cryptography              1.3                      py35_0  
curl                      7.45.0                        0  
cycler                    0.10.0                   py35_0  
cython                    0.23.4                   py35_0  
cytoolz                   0.7.5                    py35_0  
dask                      0.8.1                    py35_0  
datashape                 0.5.1                    py35_0  
decorator                 4.0.9                    py35_0  
dill                      0.2.4                    py35_0  
docutils                  0.12                     py35_0  
dynd-python               0.7.2                    py35_0  
et_xmlfile                1.0.1                    py35_0  
fastcache                 1.0.2                    py35_0  
flask                     0.10.1                   py35_1  
flask-cors                2.1.2                    py35_0  
fontconfig                2.11.1                        5  
freetype                  2.5.5                         0  
gencore-app               1.0                       <pip>
gevent                    1.1.0                    py35_0  
greenlet                  0.4.9                    py35_0  
h5py                      2.5.0               np110py35_4  
hdf5                      1.8.15.1                      2  
heapdict                  1.0.0                    py35_0  
hmm-api                   1.0                       <pip>
idna                      2.0                      py35_0  
ipykernel                 4.3.1                    py35_0  
ipython                   5.1.0                    py35_0  
ipython_genutils          0.1.0                    py35_0  
ipywidgets                4.1.1                    py35_0  
itsdangerous              0.24                     py35_0  
jbig                      2.1                           0  
jdcal                     1.2                      py35_0  
jedi                      0.9.0                    py35_0  
gencore_annotation_docs   1.0                           0    jerowe
jinja2                    2.8                      py35_0  
jpeg                      8d                            0  
jsonschema                2.4.0                    py35_0  
jupyter                   1.0.0                    py35_2  
jupyter_client            4.2.2                    py35_0  
jupyter_console           4.1.1                    py35_0  
jupyter_core              4.1.0                    py35_0  
libdynd                   0.7.2                         0  
libffi                    3.0.13                        0  
libgfortran               3.0                           0  
libpng                    1.6.17                        0  
libsodium                 1.0.3                         0  
libtiff                   4.0.6                         2  
libxml2                   2.9.2                         0  
libxslt                   1.1.28                        0  
llvmlite                  0.9.0                    py35_0  
locket                    0.2.0                    py35_0  
lxml                      3.6.0                    py35_0  
markupsafe                0.23                     py35_0  
matplotlib                1.5.1               np110py35_0  
mistune                   0.7.2                    py35_0  
mkl                       11.3.1                        0  
mkl-service               1.1.2                    py35_0  
mpmath                    0.19                     py35_0  
msgpack-python            0.4.7                    py35_0  
multipledispatch          0.4.8                    py35_0  
nbconvert                 4.1.0                    py35_0  
nbformat                  4.0.1                    py35_0  
networkx                  1.11                     py35_0  
nltk                      3.2                      py35_0  
nose                      1.3.7                    py35_0  
notebook                  4.1.0                    py35_1  
numba                     0.24.0              np110py35_0  
numexpr                   2.5                 np110py35_0  
numpy                     1.10.4                   py35_1  
odo                       0.4.2                    py35_0  
openpyxl                  2.3.2                    py35_0  
openssl                   1.0.2g                        0  
pandas                    0.18.0              np110py35_0  
partd                     0.3.2                    py35_1  
patchelf                  0.8                           0  
path.py                   8.1.2                    py35_1  
patsy                     0.4.0               np110py35_0  
pep8                      1.7.0                    py35_0  
pexpect                   4.0.1                    py35_0  
pickleshare               0.5                      py35_0  
pillow                    3.1.1                    py35_0  
pip                       8.1.1                    py35_1  
ply                       3.8                      py35_0  
prompt_toolkit            1.0.3                    py35_0  
psutil                    4.1.0                    py35_0  
ptyprocess                0.5                      py35_0  
py                        1.4.31                   py35_0  
pyasn1                    0.1.9                    py35_0  
pycosat                   0.6.1                    py35_0  
pycparser                 2.14                     py35_0  
pycrypto                  2.6.1                    py35_0  
pycurl                    7.19.5.3                 py35_0  
pyflakes                  1.1.0                    py35_0  
pygments                  2.1.1                    py35_0  
pyopenssl                 0.15.1                   py35_2  
pyparsing                 2.0.3                    py35_0  
pyqt                      4.11.4                   py35_1  
pytables                  3.2.2               np110py35_1  
pytest                    2.8.5                    py35_0  
python                    3.5.2                         0  
python-dateutil           2.5.1                    py35_0  
pytz                      2016.2                   py35_0  
pyyaml                    3.11                     py35_1  
pyzmq                     15.2.0                   py35_0  
qt                        4.8.7                         1  
qtawesome                 0.3.2                    py35_0  
qtconsole                 4.2.0                    py35_0  
qtpy                      1.0                      py35_0  
readline                  6.2                           2  
redis                     2.6.9                         0  
redis-py                  2.10.3                   py35_0  
reportlab                 3.3.0                    py35_0  
requests                  2.9.1                    py35_0  
rope                      0.9.4                    py35_1  
ruamel_yaml               0.11.14                  py35_0  
scikit-image              0.12.3              np110py35_0  
scikit-learn              0.17.1              np110py35_0  
scipy                     0.17.0              np110py35_2  
setuptools                20.3                     py35_0  
simplegeneric             0.8.1                    py35_0  
singledispatch            3.4.0.3                  py35_0  
sip                       4.16.9                   py35_0  
six                       1.10.0                   py35_0  
snowballstemmer           1.2.1                    py35_0  
sockjs-tornado            1.0.1                    py35_0  
sphinx                    1.3.5                    py35_0  
sphinx_rtd_theme          0.1.9                    py35_0  
spyder                    2.3.8                    py35_1  
sqlalchemy                1.0.12                   py35_0  
sqlite                    3.13.0                        0  
statsmodels               0.6.1               np110py35_0  
sympy                     1.0                      py35_0  
terminado                 0.5                      py35_1  
tk                        8.5.18                        0  
toolz                     0.7.4                    py35_0  
tornado                   4.3                      py35_0  
traitlets                 4.2.1                    py35_0  
unicodecsv                0.14.1                   py35_0  
util-linux                2.21                          0  
wcwidth                   0.1.7                    py35_0  
werkzeug                  0.11.4                   py35_0  
wheel                     0.29.0                   py35_0  
xlrd                      0.9.4                    py35_0  
xlsxwriter                0.8.4                    py35_0  
xlwt                      1.0.0                    py35_0  
xz                        5.2.2                         0  
yaml                      0.1.6                         0  
zeromq                    4.1.3                         0  
zlib                      1.2.8                         0 
```

I have channels r and bioconda in my .condarc.

**Comment by jerowe** [comment]
I've had success with prefixing the LD_LIBRARY_PATH (linux) and CFLAGS (osx) during build time to make sure its picking the right libraries. Let me try another build.

**Comment by jerowe** [comment]
Also, this is a completely separate issue, but I have a recipe I want to contribute to conda forge for [Neovim|https://neovim.io/], but it requires unzip as a build dependency. I saw there was talk of adding it to the docker container, but was then vetoed. I actually built it using the conda-forge/linux-anvil docker container by first installing unzip and uploaded it to my own conda channels. Tools like proper text editors, tmux, cmake, etc are so handy to have! I'm glad I got pointed towards this channel. ;-)

**Comment by jakirkham** [comment]
Hmm...still seeming to pick up [`libedit`]( https://travis-ci.org/conda-forge/sqlite-feedstock/builds/162092756#L521 ). I'll take a look locally and see what needs to be done.

File: recipe/build.sh [hunk 6]
```diff
@@ -8,7 +13,6 @@ fi
             --enable-tempstore \
             --enable-shared=yes \
             --disable-tcl \
-            --disable-readline \
```
Comment (jakirkham) [inline recipe/build.sh]: So there is an `--enable-readline` flag, which we should try. We can also try `--disable-libedit`.

File: recipe/build.sh [hunk 7]
```diff
@@ -1,5 +1,10 @@
 #!/bin/bash
 
+export CFLAGS="-I${PREFIX}/include ${CFLAGS}"
+
+export LDFLAGS="-L${PREFIX}/lib ${LDFLAGS}"
+export CPPFLAGS="-I${PREFIX}/include ${CPPFLAGS}"
```
Comment (jakirkham) [inline recipe/build.sh]: One of things we have tried setting in the past is `DYLD_FALLBACK_LIBRARY_PATH`. Though `sqlite` seems to be pretty stubborn. So maybe `DYLD_LIBRARY_PATH` would be better. The difference between the two is a matter of priority. There seems to be a `READLINE_LIBS` environment variable for `sqlite`. Would give it `readline` and `ncurses`. Unfortunately `readline` doesn't support `pkg-config` so this will all need to be manually done.

Comment (jerowe) [inline recipe/build.sh]: Lets try.

Comment (jerowe) [inline recipe/build.sh]: Is there a libedit in the $PREFIX/lib? Is there something extra we need for this/

Comment (jakirkham) [inline recipe/build.sh]: So `libedit` is a different library that provides a `readline`-like API, but is more limited in supported features. Think of it as a BSD `readline`. This is being picked up from `/usr/lib/libedit.3.dylib`. Mac ships with it as they prefer not to have `readline` due to the GPL.

We probably need some heavy duty path manipulation to make sure it doesn't search there. My last attempt managed to link against `libedit`, `readline`, and `ncurses`, but I'm not sure whether it was actually using symbols from the latter two or just adding them to the rpath.

**Comment by jakirkham** [comment]
I've asked some other interested/motivated parties to help out. So maybe they have some more advice. I can try to look at it again later, but have to go to a meeting soon.

Would just check the log for `libedit`. Normally when `conda-build` runs `install_name_tool` on all dynamically linked binaries and libraries.

**Comment by jerowe** [comment]
Alright. Thanks for all the help!

**Comment by jakirkham** [comment]
As per your question w.r.t. `neovim`, here is a PR ( https://github.com/conda-forge/staged-recipes/pull/1533 ) with `zip` and `unzip`. Maybe you could try cleaning those up and we could get them in so you could use them.

**Comment by jakirkham** [comment]
I think we need to edit/patch the `Makefile` for this to find our `readline`. There should be some environment variables that look like `READLINE_*`, which we want to set. At least this is what I find from some [ancient documentation]( https://www.sqlite.org/cvstrac/wiki?p=ReadLine ).

**Comment by jerowe** [comment]
@jakirkham , what can I do about  appveyor hanging?

**Comment by jakirkham** [comment]
Please [re-render]( https://github.com/conda-forge/conda-smithy#re-rendering-an-existing-feedstock ) with `1.4.6`. That *should* fix it.

File: .CI/build_all [hunk 8]
```diff
@@ -0,0 +1,28 @@
+#!/usr/bin/env bash
```
Comment (jakirkham) [inline .CI/build_all]: Why do we need all of this?

File: .travis.yml [hunk 9]
```diff
@@ -1,37 +1,36 @@
-# This file was generated automatically from conda-smithy. To update this configuration,
```
Comment (jakirkham) [inline .travis.yml]: Why are we changing this file?

File: .CI/build_all (hunk 8 above)
Comment (jerowe) [inline .CI/build_all]: I'm not sure. I just rerendered it.

**Comment by jerowe** [comment]
@jakirkham , am I even building for windows?

If I am, why am i doing such a thing?

**Comment by jerowe** [comment]
@jakirkham , I'm trying to get back to this recipe before I move on to some to some others.

I don't know what is happening with appveyor, and I'm afraid to dig to deeply into the windows world.

**Comment by ocefpaf** [comment]
@jerowe due to the difficulty in reabasing this  I sent #19. Please take a look at it and let me know if you have an idea on how to fix `OS X`.

Thanks!
